tests/retry_successful_execution.go

```diff
@@ -0,0 +1,109 @@
+	subnetAInfo := subnets[0]
+	subnetBInfo := subnets[1]
+	teleporterContractAddress := network.GetTeleporterContractAddress()
+	fundedAddress, fundedKey := network.GetFundedAccountInfo()
+
+	subnetATeleporterMessenger, err := teleportermessenger.NewTeleporterMessenger(
+		teleporterContractAddress, subnetAInfo.ChainRPCClient,
+	)
+	Expect(err).Should(BeNil())
+	subnetBTeleporterMessenger, err := teleportermessenger.NewTeleporterMessenger(
+		teleporterContractAddress, subnetBInfo.ChainRPCClient,
+	)
+	Expect(err).Should(BeNil())
+
+	//
+	// Deploy ExampleMessenger to Subnets A and B
+	//
+	ctx := context.Background()
+
+	_, subnetAExampleMessenger := localUtils.DeployExampleCrossChainMessenger(ctx, fundedAddress, fundedKey, subnetAInfo)
+	exampleMessengerContractB, subnetBExampleMessenger := localUtils.DeployExampleCrossChainMessenger(
+		ctx, fundedAddress, fundedKey, subnetBInfo,
+	)
+
+	//
+	// Call the example messenger contract on Subnet A
+	//
+	message := "Hello, world!"
+	optsA := utils.CreateTransactorOpts(ctx, subnetAInfo, fundedAddress, fundedKey)
+	tx, err := subnetAExampleMessenger.SendMessage(
+		optsA,
+		subnetBInfo.BlockchainID,
+		exampleMessengerContractB,
+		fundedAddress,
+		big.NewInt(0),
+		big.NewInt(300000),
+		message,
+	)
+	Expect(err).Should(BeNil())
+
+	// Wait for the transaction to be mined
+	receipt, err := bind.WaitMined(ctx, subnetAInfo.ChainRPCClient, tx)
+	Expect(err).Should(BeNil())
+	Expect(receipt.Status).Should(Equal(types.ReceiptStatusSuccessful))
+
+	event, err := utils.GetEventFromLogs(receipt.Logs, subnetATeleporterMessenger.ParseSendCrossChainMessage)
+	Expect(err).Should(BeNil())
+	Expect(event.DestinationChainID[:]).Should(Equal(subnetBInfo.BlockchainID[:]))
+
+	teleporterMessageID = event.Message.MessageID
+
+	//
+	// Relay the message to the destination
+	//
+
+	receipt = network.RelayMessage(ctx, receipt, subnetAInfo, subnetBInfo, false, true)
+	receiveEvent, err := utils.GetEventFromLogs(receipt.Logs, subnetBTeleporterMessenger.ParseReceiveCrossChainMessage)
+	Expect(err).Should(BeNil())
+	deliveredTeleporterMessage := receiveEvent.Message
+
+	//
+	// Check Teleporter message received on the destination
+	//
+	delivered, err := subnetBTeleporterMessenger.MessageReceived(
+		&bind.CallOpts{}, subnetAInfo.BlockchainID, teleporterMessageID,
+	)
+	Expect(err).Should(BeNil())
+	Expect(delivered).Should(BeTrue())
+
+	//
+	// Verify we received the expected string
+	//
+	_, currMessage, err := subnetBExampleMessenger.GetCurrentMessage(&bind.CallOpts{}, subnetAInfo.BlockchainID)
+	Expect(err).Should(BeNil())
+	Expect(currMessage).Should(Equal(message))
+
+	//
+	// Attempt to retry message execution, which should fail
+	//
+	optsB := utils.CreateTransactorOpts(ctx, subnetBInfo, fundedAddress, fundedKey)
+	tx, err = subnetBTeleporterMessenger.RetryMessageExecution(optsB, subnetAInfo.BlockchainID, deliveredTeleporterMessage)
+	Expect(err).Should(Not(BeNil()))
+	Expect(tx).Should(BeNil())
```

Comment:
why should the transaction be `nil`? Shouldn't we still be able to check transactipn receipt and make sure receipt status is failed?

Reply:
That's a good question. The warp message should still pass verification, so I'd expect the tx to be mined. I left a TODO and will create an issue to investigate.

Reply:
That's just general `go` convention to not return anything when you have an error. I would imagine we could check the error for some sort of relevant information.

Reply:
`RetryMessageExecution` issues a transaction to call `retryMessageExecution` on the `TeleporterMessenger` contract. I'd expect the transaction issuance to succeed (and the Go call here to return a non-nil tx), but the receipt to have a failed status after it has been mined. For example, in [SendTransactionAndWaitForAcceptance](https://github.com/ava-labs/teleporter/blob/9f4c142f9735fc8a87428e8c713776f92d469231/tests/utils/utils.go#L56) the `SendTransaction` call is successful, even if the expected receipt status is `ReceiptStatusFailed`.

Following the call to `RetryMessageExecution` eventually leads to [here](https://github.com/ava-labs/subnet-evm/blob/8beb366405008ee0f40337069e7bf124f8e23140/accounts/abi/bind/base.go#L418), which resolves to the same `ethclient.Client.SendTransaction` call as the SendTransactionAndWaitForAcceptance example I mentioned above. For some reason though, that produces an error, while the other example does not.

Reply:
Yeah, I guess it's the way they've implemented it in `subnet-evm`. If `BoundContract.transact()` ever errors, it does not return the tx. If we want the behaviour you're expecting here, we could write a wrapper `alwaysReturnTx` that takes in any function that returns `(*types.Transaction, error)`. We could then call the function, and if it returns an error, set `opts.NoSend = true`, and call the function again to get the tx. It's a little bit of work and not sure if it's worth it, but the option is there.

The other option is just filing a ticket in `subnet-evm` to request the change in `BoundContract.transact()`

Reply:
FYI the error contains the message `"execution reverted: TeleporterMessenger: message not found"`

Reply:
Nice, thanks for checking the error. That gives me confidence that the test is behaving as expected. It still might be worth investigating how we can unify these behaviors, but as long as the test is correct, that is very low priority IMO.